source/core/sudoku/cell.d

```diff
@@ -11,75 +12,162 @@ public class Cell
+	public void addConstraint(C : Constraint)(Cell[] cells...)
 	{
-		import std.algorithm : canFind;
-		if (constraint is null)
-			return null;
+		// remove null and repeated cells
+		import std.algorithm : canFind, filter, uniq;
+		import std.array : array;
 
-		C c = get!C;
-		if (c is null)
+		cells = uniq(cells.filter!(cell => cell !is null && cell != this)).array;
+		ConstraintType type = C.getStaticConstraintType;
+
+		// create a new instance if this Cell doesn't containt this ConstraintType
+		if (type !in constraints)
 		{
-			constraints ~= constraint;
-			c = constraint;
+			constraints[type] = new C();
 		}
-		else
+
+		// add and connect to non existent Cells
+		foreach (cell; cells)
 		{
-			foreach (Cell cell; constraint.cells)
+			if (!canFind(constraints[type].cells, cell))
 			{
-				if (cell != this)
-					c.add(cell);
+				constraints[type].cells ~= cell;
+				cell.addConstraint!C(this);
 			}
 		}
+	}
 
-		return c;
+
+	/** safely adds a constraint
+	 *
+	 * Adds a constraint by its type \
+	 * It filters any invalid data passed in `cells`, null values, duplicated objects,
+	 *     same Cell as the being used
+	 *
+	 * Params:
+	 *     type = ConstraintType to add
+	 *     cells = Cells to connect
+	 *
+	 * SeeAlso: void addConstraint(**C : Constraint**)(**Cell[] cells...**)
+	 */
+	public void addConstraint(ConstraintType type, Cell[] cells...)
+	{
+		Constraint.createMultiSingle(type, this, cells);
 	}
 
 
-	public void disconnect(C : Constraint)(Cell[] cells...)
+	/** Safely removes a constraint from a Cell
+	 *
+	 * It disconnects all the connected Cells from each other
+	 *
+	 * Params:
+	 *     type = ContraintType to remove
+	 *
+	 * SeeAlso: void disconnect(**ConstraintType type, **Cell[] cells...**)
+	 */
+	public void removeConstraint(ConstraintType type)
+	{
+		if (type !in constraints)
+			return;
+
+		// safely cut the connection between both Cells
+		disconnect(type, constraints[type].cells);
+	}
+
+
+	/** Safely removes a constraint from a Cell
+	 *
+	 * Params:
+	 *     C = Contraint to remove
+	 *
+	 * SeeAlso: void removeConstraint(**ConstraintType type**)
+	 */
+	public void removeConstraint(C : Constraint)()
 	{
-		import std.algorithm : each, remove;
+		removeConstraint(C.getStaticConstraintType);
+	}
+
+
+	/** Safely disconnects Cells from each other
+	 *
+	 * It filters any invalid data passed in `cells`, null values, duplicated objects,
+	 *     same Cell as the being used
+	 *
+	 * Params:
+	 *     type = ConstraintType to use
+	 *     cells = cells to disconnect
+	 */
+	public void disconnect(ConstraintType type, Cell[] cells...)
+	{
+		import std.algorithm : canFind, each, filter, remove, uniq;
+		import std.array : array;
 		import std.range : empty;
-		C constraint = get!C;
-		if (constraint is null)
+
+		cells = uniq(cells.filter!(cell => cell !is null && cell != this)).array;
+
+		// check if Cell has this constraint
+		if (type !in constraints)
```

Comment:
ditto

Reply:
This check has to be this way. Each Cell in the array the cells connected calls this function to remove the Cell which is disconnecting them.

```d
Cell cell1 = new Cell(0);
Cell cell2 = new Cell(0);
cell1.addConstraint!UniqueConstraint(cell2);
// at this point cell1 and cell2 are connected to each other

cell1.disconect!UniqueConstraint(cell2);
// during the execution
// cell1 checks if cell2 is in the connectedCells (true)
// cell1 removes cell2 from the array
// cell2.disconect!UniqueConstraint(cell1) is called
// cell2 checks if cell1 is in the connectedCells (true)
// cell1.disconect!UniqueConstraint(cell1) is called
// cell1 checks if cell2 is in the connectedCells (false)
// return
// return
// done
```

Reply:
If I add that contract to this function it'll break.